test/inductor/test_torchinductor.py

```diff
@@ -1147,36 +1159,68 @@ def flip(a, b):
 
         # Constant propagate a constant that's negative
         def flip_with_index_constant(a):
-            b = torch.arange(start=-1, end=-a.numel() - 1, step=-1, device=self.device)
+            b = torch.arange(start=-1, end=-a.numel() - 1, step=-1, device=a.device)
             return a[b]
 
         # Wrapping is constant-folded
         test(flip_with_index_constant, (a,), has_assert=False, has_wrapping=False)
 
         # Operation where we can't prove that the index is always positive or negative
         def pos_and_neg(a):
-            b = torch.arange(start=1, end=-a.numel() - 1, step=-1, device=self.device)
+            b = torch.arange(start=1, end=-a.numel() - 1, step=-1, device=a.device)
             return a[b]
 
         # It has wrapping but no assert
         test(pos_and_neg, (a,), has_assert=False, has_wrapping=True)
 
         # We currently don't do constant propagation with float constants
+        # We cannot prove this kind of asserts just with bounds. We would need
+        # to lift IndexPropagation.shape_env to be accessible in all of Inductor
         def flip_with_index(a):
             b = 1.0 * torch.arange(
-                start=-1, end=-a.numel() - 1, step=-1, device=self.device
+                start=-1, end=-a.numel() - 1, step=-1, device=a.device
```

Comment:
I really think we should be able to prove this one. We have `start + i0 * step => -1 -i0` which after wrapping becomes `s0 - 1 - i0`. Then we have the condition `s0 - 1 - i0 < s0` which if we bring over to the same side gives `-1 - i0 < 0`. which sympys basic reasoning should knows because  `i0` is nonnegative.

Where is it going wrong here? Do we need to rewrite `expr < size` as `expr - size < 0` manually?

Reply:
The issue here is that we multiply by a float and then cast back into int. We don't track that in `IndexPropagation` ATM (although we could).

This one was there to test an example of operation that is not propagated through `IndexPropagation`, so we rely exclusively on the bound analysis (for now).

Reply:
Oh I see, the test call below was bugged before as it calls the wrong function. nvm